File: Provenance/Controller/PV32XControllerViewController.m
//
//  PV32XControllerViewController.m
//  Provenance
//
//  Created by Joe Mattiello on 08/22/2016.
//  Copyright (c) 2016 Joe Mattiello. All rights reserved.
//

#import "PV32XControllerViewController.h"
#import <PicoDrive/PicodriveGameCore.h>
#import <PicoDrive/OESega32XSystemResponderClient.h>

@interface PV32XControllerViewController ()

@end

@implementation PV32XControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"A"]) {
            [button setTag:OESega32XButtonA];
        } else if ([[[button titleLabel] text] isEqualToString:@"B"]) {
            [button setTag:OESega32XButtonB];
        } else if ([[[button titleLabel] text] isEqualToString:@"C"]) {
            [button setTag:OESega32XButtonC];
        } else if ([[[button titleLabel] text] isEqualToString:@"X"]) {
            [button setTag:OESega32XButtonX];
        } else if ([[[button titleLabel] text] isEqualToString:@"Y"]) {
            [button setTag:OESega32XButtonY];
        } else if ([[[button titleLabel] text] isEqualToString:@"Z"]) {
            [button setTag:OESega32XButtonZ];
        }else if ([[[button titleLabel] text] isEqualToString:@"Mode"]) {
            [button setTag:OESega32XButtonMode];
        } else if ([[[button titleLabel] text] isEqualToString:@"Start"]) {
            [button setTag:OESega32XButtonStart];
        }
    }
    
    self.startButton.tag  = OESega32XButtonStart;
    self.selectButton.tag = OESega32XButtonMode;
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    
    [three2xCore didReleaseSega32XButton:OESega32XButtonUp forPlayer:0];
    [three2xCore didReleaseSega32XButton:OESega32XButtonDown forPlayer:0];
    [three2xCore didReleaseSega32XButton:OESega32XButtonLeft forPlayer:0];
    [three2xCore didReleaseSega32XButton:OESega32XButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [three2xCore didPushSega32XButton:OESega32XButtonUp forPlayer:0];
            [three2xCore didPushSega32XButton:OESega32XButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [three2xCore didPushSega32XButton:OESega32XButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [three2xCore didPushSega32XButton:OESega32XButtonUp forPlayer:0];
            [three2xCore didPushSega32XButton:OESega32XButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [three2xCore didPushSega32XButton:OESega32XButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [three2xCore didPushSega32XButton:OESega32XButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [three2xCore didPushSega32XButton:OESega32XButtonDown forPlayer:0];
            [three2xCore didPushSega32XButton:OESega32XButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [three2xCore didPushSega32XButton:OESega32XButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [three2xCore didPushSega32XButton:OESega32XButtonDown forPlayer:0];
            [three2xCore didPushSega32XButton:OESega32XButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    
    [three2xCore didReleaseSega32XButton:OESega32XButtonUp forPlayer:0];
    [three2xCore didReleaseSega32XButton:OESega32XButtonDown forPlayer:0];
    [three2xCore didReleaseSega32XButton:OESega32XButtonLeft forPlayer:0];
    [three2xCore didReleaseSega32XButton:OESega32XButtonRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    NSInteger tag = button.tag;
    [three2xCore didPushSega32XButton:tag forPlayer:0];
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    NSInteger tag = button.tag;
    [three2xCore didReleaseSega32XButton:tag forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    [three2xCore didPushSega32XButton:OESega32XButtonStart forPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    [three2xCore didReleaseSega32XButton:OESega32XButtonStart forPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    [three2xCore didPushSega32XButton:OESega32XButtonMode forPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    PicodriveGameCore *three2xCore = (PicodriveGameCore *)self.emulatorCore;
    [three2xCore didReleaseSega32XButton:OESega32XButtonMode forPlayer:player];
}

@end

File: Provenance/Controller/PV32XControllerViewController.swift
//  Converted to Swift 4 by Swiftify v4.1.6640 - https://objectivec2swift.com/
//
//  PV32XControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright (c) 2018 Joe Mattiello. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVSega32XButton {
        get {
            return PVSega32XButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVSega32XControllerViewController: PVControllerViewController<PVSega32XSystemResponderClient> {
// extension PVControllerViewController where ResponderClass : PVSega32XSystemResponderClient {
//extension ControllerVC where ResponderType : PVSega32XSystemResponderClient {
    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let title = button.titleLabel?.text else {
                return
            }
            if title == "A" {
                button.buttonTag = .a
            } else if title == "B" {
                button.buttonTag = .b
            } else if title == "C" {
                button.buttonTag = .c
            } else if title == "X" {
                button.buttonTag = .x
            } else if title == "Y" {
                button.buttonTag = .y
            } else if title == "Z" {
                button.buttonTag = .z
            } else if title == "Mode" {
                button.buttonTag = .mode
            } else if title == "Start" {
                button.buttonTag = .start
            }
        }

        startButton?.buttonTag = .start
        selectButton?.buttonTag = .mode
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        self.emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.up, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.left, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.right, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.down, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        vibrate()
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.start, forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.start, forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.mode, forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.mode, forPlayer: player)
    }
}
